### PR TITLE
Object store metadata Table provider

### PR DESCRIPTION
**Pull request by Jeadie**
## Changes
 - The key change is `impl TableProvider for ObjectStoreMetadataTable`. 
 - A table provider compatible with any [object_store](https://docs.rs/object_store/latest/object_store)
 - Table schema based on [`ObjectMeta`](https://docs.rs/object_store/latest/object_store/struct.ObjectMeta.html).

### Architecture Decisions
- `DataConnectors` can implement an optional new function
```rust
    async fn metadata_provider(
        &self,
        _dataset: &Dataset,
    ) -> Option<AnyErrorResult<Arc<dyn TableProvider>>> {
        None
    }
```
- New schema `metadata` (i.e. `spice.metadata.$table_name`). 

### UX
For a given dataset (currently only supported for object store based datasets), add a metadata boolean flag.
```yaml
datasets:
  - from: s3://spiceai-demo-datasets/taxi_trips/2024/yellow_tripdata_2024-01.parquet
    name: cleaned_sales_data_accelerated
    params:
      metadata: true
  - from: s3://spiceai-demo-datasets/tpch/customer/
    name: customer
    params:
      metadata: true
```

Creates a table in `spice.metadata.$name`:
```
sql> select * from metadata.cleaned_sales_data_accelerated
+--------------------------------+---------------------+----------+------------------------------------+---------+
| location                       | last_modified       | size     | e_tag                              | version |
+--------------------------------+---------------------+----------+------------------------------------+---------+
| tpch/customer/customer.parquet | 2024-03-29T21:43:32 | 13377306 | "07dda2f934b3fbe934a4b497aaa4cd98" |         |
+--------------------------------+---------------------+----------+------------------------------------+---------+
```

### CODE REVIEW

File: crates/runtime/src/dataconnector.rs [hunk 1]
```diff
@@ -348,16 +347,27 @@ pub trait ListingTableConnector: DataConnector {
...
         let store_url = self.get_object_store_url(dataset)?;
-        
+
         // TODO: For some reason, this `store` is not working. It's not setting `bucket_endpoint`.
         // Need this, store-specific, prefix parsing
-        let (_store, prefix) = object_store::parse_url(&store_url.clone()).unwrap();
-        let (_, extension) = self.get_file_format_and_extension()?;
+        let (_store, prefix) = object_store::parse_url(&store_url.clone()).boxed().context(
+            UnableToConnectInternalSnafu {
+                dataconnector: "ListingTableConnector",
```
Comment (phillipleblanc) [inline crates/runtime/src/dataconnector.rs]:
```suggestion
                dataconnector: format!("{self}"),
```

File: crates/runtime/src/dataconnector.rs [hunk 2]
```diff
@@ -375,11 +385,12 @@ pub trait ListingTableConnector: DataConnector {
...
 
-        let table = ObjectStoreMetadataTable::try_new(object_store, Some(prefix_str), Some(filename_regex))
-            .context(InternalSnafu {
-                dataconnector: "ListingTableConnector",
-                code: "LTC-MP-OSMT".to_string(), // ListingTableConnector-MetadataProvider-ObjectStoreMetadataTableTryNew
-            })?;
+        let table =
+            ObjectStoreMetadataTable::try_new(object_store, Some(prefix_str), Some(filename_regex))
+                .context(InternalSnafu {
+                    dataconnector: "ListingTableConnector",
```
Comment (phillipleblanc) [inline crates/runtime/src/dataconnector.rs]:
```suggestion
                    dataconnector: format!("{self}"),
```

**Review by phillipleblanc — DISMISSED** [review]

File: crates/runtime/src/dataconnector.rs [hunk 3]
```diff
@@ -319,6 +327,73 @@ pub trait ListingTableConnector: DataConnector {
...
+                    .to_string(),
+                filename.to_string(),
+            )
+        } else {
+            (prefix.to_string(), format!(r"^.*\{extension}$"))
+        };
+
+        let listing_store_url = ListingTableUrl::parse(store_url.clone()).boxed().context(
+            UnableToConnectInternalSnafu {
+                dataconnector: "ListingTableConnector",
```
Comment (phillipleblanc) [inline crates/runtime/src/dataconnector.rs]:
```suggestion
                dataconnector: format!("{self}"),
```

File: crates/runtime/src/dataconnector.rs (hunk 2 above)
Comment (Jeadie) [reply crates/runtime/src/dataconnector.rs]: Thank you. Before I had `where Self: Display` i couldn't use `format!("{self}")`

**Review by phillipleblanc — APPROVED** [review]

**Pull request merged**